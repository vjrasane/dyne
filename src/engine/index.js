export { default as engine } from "~/engine/core";
export { Cmd } from "~/engine/effects/command";
export { Sub } from "~/engine/effects/subscription";
export { Update, Update as Init } from "~/engine/core/update";
