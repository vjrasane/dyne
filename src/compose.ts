import { Update, Updated } from "./engine/core/update";
import Return from "./engine/effects/return";
import { isFunction, exists, always } from "./utils";
import { OpticalUpdate, Lens, Optical } from "./optical";

/**
 * Combine optical updates into a single update function
 *
 * @param opticals
 */
export const combine = <M>(opticals: OpticalUpdate<M, any>[]) => (
  msg: object,
  model: M
): Updated<M> =>
  opticals.reduce(
    <C>(updated: Return<M>, optical: OpticalUpdate<M, C>) =>
      updated.bind((m: M) => {
        // get child model from parent model
        const child = optical.lens.get(m);
        // call update with child
        const ret: Updated<C> = optical.func(msg, child);
        // if update returned nothing, use previous model
        if (!exists(ret)) {
          return Update(m);
        }
        // ensure each update value is stored as update monad
        const up: Return<C> = Update.wrap(ret);
        // map monad back to parent model and return
        return up.map((c: C) => optical.lens.set(c, m));
      }),
    // wrap model in update monad
    Update(model)
  );

/**
 *
 * Get array of optical updates based on given object structure
 *
 * @param struct
 * @param parentLens
 */
const getOpticals = <M>(
  struct: object,
  parentLens?: Lens<M, any>
): OpticalUpdate<M, any>[] => {
  // Flattened array of optical updates
  const opticals: OpticalUpdate<M, any>[] = [];

  Object.entries(struct).forEach(([field, value]) => {
    // Lens for the currently processed field
    const lens: Lens<M, any> = parentLens
      ? // If parent lens is present, map it with field lens
        parentLens.map(Lens.field(field))
      : // Otherwise use only field lens
        Lens.field(field);
    // If value is a function, assume it is an instance of update
    if (isFunction(value)) {
      // Create optical update with the field lens
      const optical: OpticalUpdate<M, any> = Optical(lens, value);
      // Push to list of optical updates
      opticals.push(optical);
    } else if (typeof value === "object") {
      // Recursively create opticals for each object field
      const childOpticals: OpticalUpdate<M, any>[] = getOpticals(value, lens);
      // Push child optical updates to flattened array
      opticals.push(...childOpticals);
    } else {
      // For any other field, create optical update with constant return value
      const optical: OpticalUpdate<M, any> = Optical(lens, always(value));
      // Push to list of optical updates
      opticals.push(optical);
    }
  });
  return opticals;
};

const compose = <M>(struct: object): Update<M> => combine(getOpticals(struct));

export default compose;
