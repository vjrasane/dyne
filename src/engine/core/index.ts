import { renderer, VirtualDom } from "../render";
import { getInitializer, Initializer, Init } from "../core/init";
import { getUpdater, Updater, Update } from "../core/update";
import { getViewer, View, Viewer } from "../core/view";
import { dispatch, queues } from "../core/dispatch";
import { Sink } from "../../utils";

const UPDATE_INTERVAL = 1; /* milliseconds */
const VIEW_INTERVAL = 10; /* milliseconds */
const RENDER_INTERVAL = 10; /* milliseconds */
const EFFECT_INTERVAL = 1; /* milliseconds */

const loop = (procedure: Sink, interval: number): void => {
  let lock = false;
  setInterval((): void => {
    if (!lock) {
      lock = true;
      procedure();
      lock = false;
    }
  }, interval);
};

type Setup<M> = {
  init: Init<M>;
  update: Update<M>;
  view: View<M>;
};

const engine = <M>({ init, update, view }: Setup<M>) => {
  const updater: Updater<M> = getUpdater(update);
  const viewer: Viewer<M> = getViewer(view);
  const initializer: Initializer<M> = getInitializer(init);

  return ({ node, flags }) => {
    let model: M = initializer(flags);

    // update loop
    loop(() => {
      if (queues.msg.length > 0) {
        model = updater(queues.msg.shift(), model);
      }
    }, UPDATE_INTERVAL);

    // view loop
    let initialRender: Boolean = false;
    let viewedModel: M;
    let virtualDom: VirtualDom;
    const render = () => {
      virtualDom = viewer(model);
      viewedModel = model;
    };

    loop(() => {
      if (!initialRender) {
        render();
        initialRender = true;
      } else if (model && model !== viewedModel) {
        render();
      }
    }, VIEW_INTERVAL);

    // render loop
    let renderedVirtualDom;
    loop(() => {
      if (virtualDom && virtualDom !== renderedVirtualDom) {
        renderer(node, virtualDom, renderedVirtualDom);
        renderedVirtualDom = virtualDom;
      }
    }, RENDER_INTERVAL);

    // effect loop
    loop(() => {
      if (queues.effect.length > 0) {
        const effect = queues.effect.shift();
        // execute effect with the message dispatcher
        effect.execute(dispatch);
      }
    }, EFFECT_INTERVAL);
  };
};

export default engine;
