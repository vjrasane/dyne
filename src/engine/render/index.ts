import { isFunction, exists } from "../utils";
import { msgEventListener } from "../effects/command";
import { isElement, DyneElement } from "../../element";
import { dispatch } from "../core/dispatch";

const isEventProp = (name: string): boolean => /^on/.test(name);

const extractEventName = (name: string): string => name.slice(2).toLowerCase();

const removeBooleanProp = ($target: HTMLElement, name: string): void => {
  $target.removeAttribute(name);
  $target[name] = false;
};

const removeProp = (
  $target: HTMLElement,
  name: string,
  value: string | boolean
) => {
  if (name === "className") {
    $target.removeAttribute("class");
  } else if (typeof value === "boolean") {
    removeBooleanProp($target, name);
  } else {
    $target.removeAttribute(name);
  }
};

const setBooleanProp = ($target: HTMLElement, name: string, value) => {
  if (value) {
    $target.setAttribute(name, value);
    $target[name] = true;
  } else {
    $target[name] = false;
  }
};

const propChanged = (firstProp: Prop, secondProp: Prop) =>
  !(
    firstProp === secondProp ||
    (isFunction(firstProp) &&
      isFunction(secondProp) &&
      firstProp.toString() === secondProp.toString())
  );

const nodeChanged = (firstNode: DomElement, secondNode: VirtualDom) =>
  !(
    firstNode === secondNode ||
    (firstNode instanceof DyneElement &&
      secondNode instanceof DyneElement &&
      firstNode.type === secondNode.type)
  );

type Prop = string | boolean | Function;

const setProp = ($target: HTMLElement, name: string, value): void => {
  if (isEventProp(name)) {
    $target.addEventListener(
      extractEventName(name),
      msgEventListener(value, dispatch)
    );
  } else if (name === "className") {
    $target.setAttribute("class", value);
  } else if (typeof value === "boolean") {
    setBooleanProp($target, name, value);
  } else {
    $target.setAttribute(name, value);
  }
};

const setProps = ($target: HTMLElement, props: object): void => {
  Object.keys(props).forEach(name => {
    setProp($target, name, props[name]);
  });
};

const updateProp = (
  $target: HTMLElement,
  name: string,
  newValue: string,
  oldValue: string
) => {
  if (!newValue) {
    removeProp($target, name, oldValue);
  } else if (!oldValue || propChanged(newValue, oldValue)) {
    setProp($target, name, newValue);
  }
};

const updateProps = (
  $target: HTMLElement,
  newProps: object,
  oldProps: object = {}
) => {
  const props = Object.assign({}, newProps, oldProps);
  Object.keys(props).forEach(name => {
    updateProp($target, name, newProps[name], oldProps[name]);
  });
};

const createElement = (node: DyneElement | string) => {
  if (node instanceof DyneElement) {
    const $el = document.createElement(node.type);
    setProps($el, node.props);
    node.children.map(createElement).forEach($el.appendChild.bind($el));
    return $el;
  }
  return document.createTextNode(node);
};

type DomElement = string | DyneElement;

export type VirtualDom = DomElement | DomElement[];

const updateElement = (
  $parent: HTMLElement,
  newNode: DomElement,
  oldNode: VirtualDom,
  index = 0
) => {
  if (!exists(oldNode)) {
    $parent.appendChild(createElement(newNode));
  } else if (!exists(newNode)) {
    $parent.removeChild($parent.childNodes[index]);
  } else if (nodeChanged(newNode, oldNode)) {
    $parent.replaceChild(createElement(newNode), $parent.childNodes[index]);
  } else if (newNode instanceof DyneElement) {
    const oldElem =
      oldNode instanceof DyneElement ? oldNode : DyneElement.empty;
    updateProps(
      <HTMLElement>$parent.childNodes[index],
      newNode.props,
      oldElem.props
    );
    const newLength = newNode.children.length;
    const oldLength = oldElem.children.length;
    for (let i = 0; i < Math.max(newLength, oldLength); i++) {
      updateElement(
        <HTMLElement>$parent.childNodes[index],
        newNode.children[i],
        oldElem.children[i],
        i
      );
    }
  }
};

export const renderer = (
  $root: HTMLElement,
  newDom: VirtualDom,
  oldDom: VirtualDom
): void => {
  if (Array.isArray(newDom)) {
    newDom.forEach((node: DyneElement, index: number) =>
      updateElement($root, node, oldDom ? oldDom[index] : null, index)
    );
  } else {
    updateElement($root, newDom, oldDom);
  }
};
