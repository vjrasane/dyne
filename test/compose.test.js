import { Cmd, Update } from "~/engine";

import compose from "~/compose";

describe("compose", () => {
  describe("model", () => {
    const update = id => (msg, model) => ({
      ...model,
      id,
      msg,
      status: "updated"
    });

    it("updates single field correctly", () => {
      const composed = compose({
        child: update("single")
      });

      const model = { child: { field: "value" } };
      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("updates embedded field correctly", () => {
      const composed = compose({
        firstChild: {
          secondChild: {
            thirdChild: update("embedded")
          }
        }
      });

      const model = {
        firstChild: {
          secondChild: {
            thirdChild: {
              field: "value"
            }
          }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("updates array of fields somewhat correctly", () => {
      const composed = compose({
        array: [update(0), update(1), update(2)]
      });

      const model = {
        array: [{ field: 0 }, { field: 1 }, { field: 2 }]
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("keeps constant value", () => {
      const composed = compose({
        constant: "value"
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("updates missing field", () => {
      const composed = compose({
        missing: {
          field: update("missing")
        }
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("updates complex structure correctly", () => {
      const composed = compose({
        firstChild: update("first"),
        secondChild: update("second"),
        thirdChild: update("third"),
        embeddedChild: {
          embedded: update("embedded"),
          constant: "embedded"
        },
        constant: "value"
      });

      const model = {
        firstChild: { field: "first" },
        secondChild: { field: "second" },
        thirdChild: { field: "third" },
        embeddedChild: {
          embedded: { field: "embedded" }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });
  });

  describe("command", () => {
    const update = () => Cmd(() => "result");

    it("returns single command correctly", () => {
      const composed = compose({
        child: update
      });

      const model = { child: { field: "value" } };
      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("returns embedded command correctly", () => {
      const composed = compose({
        firstChild: {
          secondChild: {
            thirdChild: update
          }
        }
      });

      const model = {
        firstChild: {
          secondChild: {
            thirdChild: {
              field: "value"
            }
          }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("returns array of commands somewhat correctly", () => {
      const composed = compose({
        array: [update, update, update]
      });

      const model = {
        array: [{ field: 0 }, { field: 1 }, { field: 2 }]
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("keeps constant value", () => {
      const composed = compose({
        constant: "value"
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("updates missing field", () => {
      const composed = compose({
        missing: {
          field: update
        }
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("returns complex command structure correctly", () => {
      const composed = compose({
        firstChild: update,
        secondChild: update,
        thirdChild: update,
        embeddedChild: {
          embedded: update,
          constant: "embedded"
        },
        constant: "value"
      });

      const model = {
        firstChild: { field: "first" },
        secondChild: { field: "second" },
        thirdChild: { field: "third" },
        embeddedChild: {
          embedded: { field: "embedded" }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });
  });

  describe("update", () => {
    const update = id => (msg, model) =>
      Update(
        {
          ...model,
          id,
          msg,
          status: "updated"
        },
        Cmd(() => "result")
      );

    it("updates single field and returns command correctly", () => {
      const composed = compose({
        child: update("single")
      });

      const model = { child: { field: "value" } };
      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("updates embedded field and returns command correctly", () => {
      const composed = compose({
        firstChild: {
          secondChild: {
            thirdChild: update("embedded")
          }
        }
      });

      const model = {
        firstChild: {
          secondChild: {
            thirdChild: {
              field: "value"
            }
          }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("updates array of fields and returns commands somewhat correctly", () => {
      const composed = compose({
        array: [update(0), update(1), update(2)]
      });

      const model = {
        array: [{ field: 0 }, { field: 1 }, { field: 2 }]
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("keeps constant value", () => {
      const composed = compose({
        constant: "value"
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("updates missing field", () => {
      const composed = compose({
        missing: {
          field: update("missing")
        }
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("updates complex field structure and returns commands correctly", () => {
      const composed = compose({
        firstChild: update("first"),
        secondChild: update("second"),
        thirdChild: update("third"),
        embeddedChild: {
          embedded: update("embedded"),
          constant: "embedded"
        },
        constant: "value"
      });

      const model = {
        firstChild: { field: "first" },
        secondChild: { field: "second" },
        thirdChild: { field: "third" },
        embeddedChild: {
          embedded: { field: "embedded" }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });
  });

  describe("nothing", () => {
    const update = () => {};

    it("keeps single field", () => {
      const composed = compose({
        child: update
      });

      const model = { child: { field: "value" } };
      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("keeps embedded field", () => {
      const composed = compose({
        firstChild: {
          secondChild: {
            thirdChild: update
          }
        }
      });

      const model = {
        firstChild: {
          secondChild: {
            thirdChild: {
              field: "value"
            }
          }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("keeps array of fields somewhat", () => {
      const composed = compose({
        array: [update, update, update]
      });

      const model = {
        array: [{ field: 0 }, { field: 1 }, { field: 2 }]
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });

    it("keeps constant value", () => {
      const composed = compose({
        constant: "value"
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("keeps missing field", () => {
      const composed = compose({
        missing: {
          field: update
        }
      });

      const updated = composed("message", {});
      expect(updated).toMatchSnapshot();
    });

    it("keeps complex field structure", () => {
      const composed = compose({
        firstChild: update,
        secondChild: update,
        thirdChild: update,
        embeddedChild: {
          embedded: update,
          constant: "embedded"
        },
        constant: "value"
      });

      const model = {
        firstChild: { field: "first" },
        secondChild: { field: "second" },
        thirdChild: { field: "third" },
        embeddedChild: {
          embedded: { field: "embedded" }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });
  });

  describe("combination", () => {
    const updateModel = id => (msg, model) => ({
      ...model,
      id,
      msg,
      status: "updated"
    });

    const updateCmd = () => Cmd(() => "result");

    const updateBoth = id => (msg, model) =>
      Update(updateModel(id)(msg, model), updateCmd());

    const updateNothing = () => {};

    it("updates complex field structure and returns commands correctly", () => {
      const composed = compose({
        model: updateModel("model"),
        cmd: updateCmd,
        both: updateBoth("both"),
        nothing: updateNothing,
        constant: "value",
        embedded: {
          model: updateModel("model"),
          cmd: updateCmd,
          both: updateBoth("both"),
          nothing: updateNothing,
          constant: "embedded"
        }
      });

      const model = {
        model: { field: "model" },
        cmd: { field: "cmd" },
        both: { field: "both" },
        nothing: { field: "nothing" },
        embedded: {
          model: { field: "model" },
          cmd: { field: "cmd" },
          both: { field: "both" },
          nothing: { field: "nothing" }
        }
      };

      const updated = composed("message", model);
      expect(updated).toMatchSnapshot();
    });
  });
});
