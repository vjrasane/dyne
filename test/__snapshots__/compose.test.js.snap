// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose combination updates complex field structure and returns commands correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "both": Object {
      "field": "both",
      "id": "both",
      "msg": "message",
      "status": "updated",
    },
    "cmd": null,
    "constant": "value",
    "embedded": Object {
      "both": Object {
        "field": "both",
        "id": "both",
        "msg": "message",
        "status": "updated",
      },
      "cmd": null,
      "constant": "embedded",
      "model": Object {
        "field": "model",
        "id": "model",
        "msg": "message",
        "status": "updated",
      },
      "nothing": Object {
        "field": "nothing",
      },
    },
    "model": Object {
      "field": "model",
      "id": "model",
      "msg": "message",
      "status": "updated",
    },
    "nothing": Object {
      "field": "nothing",
    },
  },
}
`;

exports[`compose command keeps constant value 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "constant": "value",
  },
}
`;

exports[`compose command returns array of commands somewhat correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "array": Object {
      "0": null,
      "1": null,
      "2": null,
    },
  },
}
`;

exports[`compose command returns complex command structure correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "constant": "value",
    "embeddedChild": Object {
      "constant": "embedded",
      "embedded": null,
    },
    "firstChild": null,
    "secondChild": null,
    "thirdChild": null,
  },
}
`;

exports[`compose command returns embedded command correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": null,
      },
    },
  },
}
`;

exports[`compose command returns single command correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "child": null,
  },
}
`;

exports[`compose command updates missing field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "missing": Object {
      "field": null,
    },
  },
}
`;

exports[`compose model keeps constant value 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "constant": "value",
  },
}
`;

exports[`compose model updates array of fields somewhat correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "array": Object {
      "0": Object {
        "field": 0,
        "id": 0,
        "msg": "message",
        "status": "updated",
      },
      "1": Object {
        "field": 1,
        "id": 1,
        "msg": "message",
        "status": "updated",
      },
      "2": Object {
        "field": 2,
        "id": 2,
        "msg": "message",
        "status": "updated",
      },
    },
  },
}
`;

exports[`compose model updates complex structure correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "constant": "value",
    "embeddedChild": Object {
      "constant": "embedded",
      "embedded": Object {
        "field": "embedded",
        "id": "embedded",
        "msg": "message",
        "status": "updated",
      },
    },
    "firstChild": Object {
      "field": "first",
      "id": "first",
      "msg": "message",
      "status": "updated",
    },
    "secondChild": Object {
      "field": "second",
      "id": "second",
      "msg": "message",
      "status": "updated",
    },
    "thirdChild": Object {
      "field": "third",
      "id": "third",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`compose model updates embedded field correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": Object {
          "field": "value",
          "id": "embedded",
          "msg": "message",
          "status": "updated",
        },
      },
    },
  },
}
`;

exports[`compose model updates missing field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "missing": Object {
      "field": Object {
        "id": "missing",
        "msg": "message",
        "status": "updated",
      },
    },
  },
}
`;

exports[`compose model updates single field correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "child": Object {
      "field": "value",
      "id": "single",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`compose nothing keeps array of fields somewhat 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "array": Array [
      Object {
        "field": 0,
      },
      Object {
        "field": 1,
      },
      Object {
        "field": 2,
      },
    ],
  },
}
`;

exports[`compose nothing keeps complex field structure 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "constant": "value",
    "embeddedChild": Object {
      "constant": "embedded",
      "embedded": Object {
        "field": "embedded",
      },
    },
    "firstChild": Object {
      "field": "first",
    },
    "secondChild": Object {
      "field": "second",
    },
    "thirdChild": Object {
      "field": "third",
    },
  },
}
`;

exports[`compose nothing keeps constant value 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "constant": "value",
  },
}
`;

exports[`compose nothing keeps embedded field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": Object {
          "field": "value",
        },
      },
    },
  },
}
`;

exports[`compose nothing keeps missing field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {},
}
`;

exports[`compose nothing keeps single field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "child": Object {
      "field": "value",
    },
  },
}
`;

exports[`compose update keeps constant value 1`] = `
Return {
  "bind": [Function],
  "effects": Array [],
  "map": [Function],
  "model": Object {
    "constant": "value",
  },
}
`;

exports[`compose update updates array of fields and returns commands somewhat correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "array": Object {
      "0": Object {
        "field": 0,
        "id": 0,
        "msg": "message",
        "status": "updated",
      },
      "1": Object {
        "field": 1,
        "id": 1,
        "msg": "message",
        "status": "updated",
      },
      "2": Object {
        "field": 2,
        "id": 2,
        "msg": "message",
        "status": "updated",
      },
    },
  },
}
`;

exports[`compose update updates complex field structure and returns commands correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "constant": "value",
    "embeddedChild": Object {
      "constant": "embedded",
      "embedded": Object {
        "field": "embedded",
        "id": "embedded",
        "msg": "message",
        "status": "updated",
      },
    },
    "firstChild": Object {
      "field": "first",
      "id": "first",
      "msg": "message",
      "status": "updated",
    },
    "secondChild": Object {
      "field": "second",
      "id": "second",
      "msg": "message",
      "status": "updated",
    },
    "thirdChild": Object {
      "field": "third",
      "id": "third",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;

exports[`compose update updates embedded field and returns command correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "firstChild": Object {
      "secondChild": Object {
        "thirdChild": Object {
          "field": "value",
          "id": "embedded",
          "msg": "message",
          "status": "updated",
        },
      },
    },
  },
}
`;

exports[`compose update updates missing field 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "missing": Object {
      "field": Object {
        "id": "missing",
        "msg": "message",
        "status": "updated",
      },
    },
  },
}
`;

exports[`compose update updates single field and returns command correctly 1`] = `
Return {
  "bind": [Function],
  "effects": Array [
    Command {
      "effect": [Function],
      "failure": undefined,
      "success": undefined,
    },
  ],
  "map": [Function],
  "model": Object {
    "child": Object {
      "field": "value",
      "id": "single",
      "msg": "message",
      "status": "updated",
    },
  },
}
`;
